subprojects/core/src/main/java/org/gradle/process/internal/DefaultExecActionFactory.java

```diff
@@ -406,11 +409,31 @@ public void executionFinished(ExecHandle execHandle, ExecResult execResult) {
         }
     }
 
-    private static class ImmutableJavaForkOptions implements JavaForkOptionsInternal {
+    static class ImmutableJavaForkOptions implements JavaForkOptionsInternal {
         private final JavaForkOptionsInternal delegate;
-
-        public ImmutableJavaForkOptions(JavaForkOptionsInternal delegate) {
+        private final Property<String> defaultCharacterEncoding;
+        private final Property<String> minHeapSize;
+        private final Property<String> maxHeapSize;
+        private final List<String> jvmArgs;
+        private final ConfigurableFileCollection bootstrapClasspath;
+        private final Property<Boolean> enableAssertions;
+        private final Property<Boolean> debug;
+
+        public ImmutableJavaForkOptions(ObjectFactory objectFactory, FileCollectionFactory fileCollectionFactory, JavaForkOptionsInternal delegate) {
             this.delegate = delegate;
+            this.defaultCharacterEncoding = objectFactory.property(String.class).convention(delegate.getDefaultCharacterEncoding());
+            this.defaultCharacterEncoding.disallowChanges();
+            this.minHeapSize = objectFactory.property(String.class).convention(delegate.getMinHeapSize());
+            this.minHeapSize.disallowChanges();
+            this.maxHeapSize = objectFactory.property(String.class).convention(delegate.getMaxHeapSize());
+            this.maxHeapSize.disallowChanges();
+            this.jvmArgs = ImmutableList.copyOf(delegate.getJvmArgs());
+            this.bootstrapClasspath = fileCollectionFactory.configurableFiles().from(delegate.getBootstrapClasspath());
+            this.bootstrapClasspath.disallowChanges();
+            this.enableAssertions = objectFactory.property(Boolean.class).convention(delegate.getEnableAssertions());
+            this.enableAssertions.disallowChanges();
+            this.debug = objectFactory.property(Boolean.class).convention(delegate.getDebug());
+            this.debug.disallowChanges();
         }
```

Comment:
:x: This now looks more like a `ReadOnlyJavaForkOptions`, no? Like, the underlying object can still change, which _will_ make this object change through the mapped values. 

There seems to be only one place where this is actually used, to create a `DaemonForkOptions`. We should understand what is needed there and adopt the code accordingly: either rename this, or resolve eagerly and create an actual immutable copy.

Reply:
Note that current implementation is also not immutable, since getters delegate to delegate. And values of delegate can change. 

There is a small difference that here also return values from getter can later change.

But yeah, from the usage it seems we need just resolved values in the end.

Reply:
I think we can fix that and cherry-pick change to master directly to lessen the size of changes here

Reply:
Fixed via https://github.com/gradle/gradle/pull/30936